### ISSUE TITLE
During creation of HMM entries, HMM entries are deleted

### ISSUE THREAD

**Issue by ekiefl** [merenlab/anvio#1369]
Here is a minimal reproducible example. Save this fasta as test.fa:

```
>test
TCCCAATAGGATTCAAACCTATAACCTTTTGATCCGTAGTCAAATGCTCTATTCAGTTGAGCTATGGGACCAGTTCAGGA
CTTAGTCCTGTAACTTATTTAGTCTCTGCGTTAGCAGCACCTTCTTCTTTAAGAATCACGAGCTTCTTCTCCTTGATTCT
AGCCTTCTTACCAGAAAGTCCTCTGAGGTAGAAGATTCTTGCTCTACGTACGTGACCAACCTTGTTAACTTCGATAGATT
CGATAGATGGTGACTGGAATGGGAAAATTCTCTCAACACCTACACCGCCTGAAATCTTACGAACTGTGAAAGTCTTCGTG
AATGGTGATGATCCGCTGATCTGGATAACTACACCTCTAAACTTCTGAAGTCTTTCCTTGTCACCCTCTTTAATCTTGAG
GCTTACGGTGATTGTGTCACCTGCCTTGAATTTTGGATAACTTGCAGCTTTCTCGTTTGTAAAAGACTGCTCTACTAATT
TTATCAAATCCATTTCTAAATCACTTTTTAAGCGCAACGTTACTCAAATACTCAACGATTGTAAGAGGTTGCTTTGTTAA
TGGGATTGCAAAGGTACAACCTTTTTTTTAATCTGCAAGCTTTTTTAGAAATTTTCTTTCATTTTTTCGAAAAGATTTCG
ATCCTCGTTGCTGAGCACAGGTGTGAACGTGTCACTTGACCTCATGCCTTCGAGGGCACTCTTCTCGTTGCTCTTCAGCT
TCTTAGGTATCCAGACAAGGTATTTTACATATAGGTCTCCGGTGCCTGAACCATATCCTTTCACTGCAGGAAGTCCACGT
CCTCTCAGACGTACGATTGTTCCGGACTGTGTTCCAGGTTCTACCTTTATCTTATAGTTACCATCTAAGCAAGGTACTGA
AACTTCTCCTCCGAGGATAGCGTCCGTAACAGGTATTATCTTGGTATAGAACAGGTTCTTGCCCTCTCTCTTCAGTGATT
CGTGTGGAATCTCCTCGATAACGACGAGCAAGTCTCCGTTTATGCCGCCGGCTTTTGCTGAATGTCCTTCACCCTTTAGG
GTAAGCTGCATTCCGTCTTCCACGCCCGCAGGAATGTTTACCTTTACAGTAACTCTCTTTCTTACAAGTCCGCTACCGCC
GCAAGTGTGACATGGGTTCGTGACGATTTTTCCTTCACCGCCGCAACGTGAACAAGTTGTGTATGTTACCGTCTGTCCGA
AGATGCTGTTTGCCACTCTCTGTTCCTGGCCGGTGCCGTGGCATACCGGACAAGTCTTCACATCAGAAGCATTGACAGTA
CCCTTTCCGGCGCAGTCGGGGCATTTCTCATAGCGGTCGATGGTGACTTCCTTGGTCACGCCCTTGGCGATTTCTTCGAG
CGTCAGCCTTACGCGGGTACGGATGTCGCGTCCGCGCATTACTCTAGGCTGGGCCTGGCCGCGTTGTCCGCCGCCGAACC
CTCCAAAGCCGCCGAAGCCGGAAAATCCGCCGAATCCGCCACCTCCGAACAGGTCGTTCAGGATGTCGTTCAGGTTTCCG
AATCCGCCGCTGAAGTCAGGTCCCGGGTTTCCGCCCAAGCCGGCCTGTCCGAAACGGTCATATTTTGCACGTTTGTCAGG
GTCGCTCAGGACTTCATAAGCCTCAGATGCTTCCTTGAATTTCTCTTCTGCGGTTTTCTTCTCTTCGTCAGTTCCGTTTA
CCCACCTGTCAGGATGCCACTTGAGAGCCATCTTCTTGAAGGCTGACTTTATGTCATCCGCACTGGCGTTCTTATCTATG
CCGAGTACTTCGTAATAATCTCTTTTCTCTGCCATAATTTCCTCCTTCGGAATATCGATCTGCTATATTCCGACCACTAC
CTTAGCGTAGCGCAATATCTTGCCGTTGAACAGGTAACCTGTTTCCACGACATCGTATACCAATCCTTTTTTCTCTTTCT
CTGCTACTGGAACC
```

Then run

`rm CONTIGS.db; anvi-gen-contigs-database -f test.fa; anvi-run-hmms -c CONTIGS.db`

At the end we see

```
CITATION
===============================================
Anvi'o will use 'tRNAScan-SE' by Chan and Lowe (doi:10.1007/978-1-4939-9173-0_1)
to identify tRNA sequences in your data. When you publish your findings, please
do not forget to properly credit their work.

Num tRNA genes parsed ........................: 1
Num tRNA genes with introns ..................: 0

Num tRNA genes recovered .....................: 1
Gene calls added to db .......................: 1 (from source "Transfer_RNAs")

WARNING
===============================================
1 gene caller ids that were added via the HMM source have been removed from
"gene_amino_acid_sequences, genes_taxonomy"
```

The gene is added and then immediately removed. The reason is because `TablesForHMMHits.append` explictly calls `TablesForHMMHits.remove_source`. This wasn't an issue before because up until 5 minutes ago `remove_source` didn't remove entries from gene_amino_acid_sequences and genes_taxonomy.

@meren can you please try and fix this? I am having a tough time parsing the `append` logic

**Comment by meren** [comment]
I'm on this. thank you.

**Comment by meren** [comment]
Why is this happening is a bit complicated and goes back to the fact that HMMs that operate on contigs (rather than genes) require their hits to be added to the contigs database as new gene calls for seamless integration.

When we scan tRNAs (or run rRNA HMMs), we first figure out locations of hits on contigs. We then run a single additional function to interject these new genes using the tRNAs or rRNAs as source (instead of our default gene caller) to the database and update the search results dictionary with these new gene calls using,

```
search_results_dict = tables_for_hmm_hits.add_new_gene_calls_to_contigs_db_and_update_serach_results_dict(self.kind_of_search,
                                                                                                          search_results_dict,
                                                                                                          skip_amino_acid_sequences=True)
```

THEN, we call `append` to add the new HMM hits (which now have their own gene calls that didn't exist before) to the relevant tables in the database using,

```
tables_for_hmm_hits.append(self.source_name,
                           self.reference,
                           self.kind_of_search,
                           self.domain,
                           self.all_genes_searched_against,
                           search_results_dict)
```

The purpose of `append` here is to replace the contents of previous HMM runs for a given source. This function operated only on tables that were going to be updated by `append`, so everything worked beautifully and you could run the same HMM source so many times without having to delete previous ones.

But because our new `remove_source` since #1368 includes the removal of associated gene caller ids that are NOT added back by append poopies hit the fan.

So. This is the problem. I will now try to find a solution that covers this.

**Comment by ekiefl** [comment]
That makes sense to me. Sorry for buggering this up. I did not pay close enough attention during my tests before merge.

**Comment by meren** [comment]
It took much longer to address this, but I wanted to fix this once and for all. I decided to remove the `remove_source` call from `append`, which required a much cleaner design.

A PR is coming in a minute.

**Comment by ekiefl** [comment]
Fixed in #1371
